sql: lift descriptor State enum to to top level

thoszhang opened the pull request:
This commit moves the `State` protobuf enum out of the `TableDescriptor`
namespace to the top level and renames it to `DescriptorState`, and
replaces all the existing `State` enums for the other descriptors with
`DescriptorState`. This is prompted by the fact that type, schema, and
database descriptors also share the dropped and offline states with
tables.

Note that this is a backward-incompatible protobuf change from earlier
20.2 alphas. Due to the existing ordering of enum values, non-table
descriptors on disk in the dropped state, on unmarshalling, will be in
the adding state, and will essentially act as unresolvable descriptors
that will never be properly dropped without intervention. Descriptors in
the public state will not undergo any changes. We expect the unwanted
dropped -> adding conversion to be a rare occurence; users would have to
upgrade to a beta right concurrently with dropping a non-table
descriptor.

Release note (backward-incompatible change): A change to the on-disk
representation for user-defined schemas, enums, and databases was made
that is backward-incompatible with 20.2 alpha releases. Only schemas,
enums, and databases in the middle of being dropped at the time of the
upgrade are affected, and the upgrade may result in the drop never
running to completion. It's advisable to avoid dropping schemas, enums
and databases if upgrading from a 20.2 alpha release to a 20.2 beta
release.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53387)
<!-- Reviewable:end -->

Commit:
sql: lift descriptor State protobuf enum to to top level

This commit moves the `State` protobuf enum out of the `TableDescriptor`
namespace to the top level and renames it to `DescriptorState`, and
replaces all the existing `State` enums for the other descriptors with
`DescriptorState`. This is prompted by the fact that type, schema, and
database descriptors also share the dropped and offline states with
tables.

Note that this is a backward-incompatible protobuf change from earlier
20.2 alphas. Due to the existing ordering of enum values, non-table
descriptors on disk in the dropped state, on unmarshalling, will be in
the adding state, and will essentially act as unresolvable descriptors
that will never be properly dropped without intervention. Descriptors in
the public state will not undergo any changes. We expect the unwanted
dropped -> adding conversion to be a rare occurence; users would have to
upgrade to a beta right concurrently with dropping a non-table
descriptor.

Release note (backward-incompatible change): A change to the on-disk
representation for user-defined schemas, enums, and databases was made
that is backward-incompatible with 20.2 alpha releases. Only schemas,
enums, and databases in the middle of being dropped at the time of the
upgrade are affected, and the upgrade may result in the drop never
running to completion. It's advisable to avoid dropping schemas, enums
and databases if upgrading from a 20.2 alpha release to a 20.2 beta
release.

lucy-zhang commented:
The hope is specifically to get this into 20.2 beta 1 to avoid the disruptiveness of even slightly backward-incompatible protobuf changes between betas. This PR does nothing except renaming; I figure we can do other code changes later.

Keeping @dt tagged because I think bulkio has the most vested interest in the descriptor states aside from us. #53290 has a discussion of why all the descriptors need offline states during RESTORE.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I think this is reasonable enough at this phase. My sense is we should sprint towards virtualizing descriptors and all of their interactions as once that's the case, these sorts of kludges in the encoding won't stick out as much and we won't be so tempted to do this sort of unprincipled thing. 

> Reviewed 29 of 29 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53387#-:-MFYqo-mESgS8G0T8sV1:b-7wj5gv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @rohany)



<!-- Sent from Reviewable.io -->

lucy-zhang commented:
TFTRs

bors r+
